### PR TITLE
Loot Git SSH Keys

### PR DESCRIPTION
**Pull request by wdahlenburg**
This post exploitation module is used to test ssh keys on a compromised server for Git access. The primary use case is for testing access to GitHub, but similar solutions are feasible.

An SSH key can allow private repositories to be downloaded. This may contain sensitive information that may have been unavailable otherwise.

An SSH key can also be used to modify and upload code on behalf of the user. This could be used to insert a backdoor into existing code. 

This module will attempt to utilize the existing private keys on a host and authenticate against some Git server (GitHub by default). If successful, it will store the private key as loot and print out the Git user that was accessed.

This module attempts to be respectful of existing .ssh/config files by backing them up and removing evidence of the temporary .ssh/config file. It also does not append the Git server to the .ssh/known_hosts file, so evidence of this module being run is limited.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] Obtain a session on linux (root or user account)
- [ ] `use post/linux/gather/git_keys`
- [ ] `set session 1`
- [ ] **Set** GITSERVER to github.mycompany.com if needed, else github.com is default
- [ ] `run`
- [ ] **Verify** a session on a user account will only test the user's ssh keys
- [ ] **Verify** a session on a root account will test every user's ssh keys

### CODE REVIEW

**Comment by dwelch-r7** [comment]
hey @wdahlenburg it looks like we already have a post module for gathering ssh keys and storing them in loot (post/multi/gather/ssh_creds), wondering if it might make more sense to have this as an aux module that can use the keys stored in loot and try them against different git servers

**Comment by wdahlenburg** [comment]
@dwelch-r7 The purpose of this module is less about storing the ssh keys than determining further access with Git. This could definitely be taken out of the module. It was added for convenience of just pulling the working keys.

As for an aux module, that would definitely be doable. The one difference is that the outbound ssh connection would be routed through your machine versus the one with a session. After GitHub's initial public key verification, it doesn't seem to log what IP the SSH connection was made from. 

What are your thoughts?

**Comment by dwelch-r7** [comment]
Well if it was changed to be an aux module you could always use the proxies option if the request needed to come from the targets machine for whatever reason

is there anyone else that has strong feelings one way or the other?

**Comment by acammack-r7** [comment]
This is nifty idea, though I agree that this should probably be part of another module. I think in an ideal world we could do this with our `ssh_login` aux module, but it currently doesn't use stored keys and it's a bit fiddly to get working through a session. If you or @dwelch-r7 want to take a stab at that it would be pretty cool. An easier lift that still reduces extra code would be to add a test like this `post/multi/gather/ssh_creds`.

**Comment by wdahlenburg** [comment]
I like the idea of an aux module better. Reusing stored keys has more portability across different operating systems than attempting to find the keys on just a linux host. I'll start working on this and add documentation.

I'll look into adding functionality of using stored keys with the ssh_login module, but that should be tracked separately.

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  # OPTPath will revert to pwd when set back to ""
+  def key_dir
+    datastore['KEY_DIR'] != `pwd`.strip ? datastore['KEY_DIR'] : ""
+  end
+
+  def key_file
+    datastore['KEY_FILE'] != `pwd`.strip ? datastore['KEY_FILE'] : ""
+  end
+
+  def check_key_for_passphrase(file)
```
Comment (bcoles) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]:
```suggestion
  def has_passphrase?(file)
```

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    end
+    keys = []
+    this_key = []
+    in_key = false
+    keyfile.split("\n").each do |line|
+      in_key = true if (line =~ /^-----BEGIN ([RD]SA|OPENSSH) PRIVATE KEY-----/)
+      this_key << line if in_key
+      if (line =~ /^-----END ([RD]SA|OPENSSH) PRIVATE KEY-----/)
+        in_key = false
+        keys << file unless check_key_for_passphrase(file)
```
Comment (bcoles) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]:
```suggestion
        keys << file unless has_passphrase?(file)
```

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    end
+    return keys
+  end
+
+  def provide_user(output)
+    vprint_status("SSH Test: #{output}")
+    if output.include? 'successfully authenticated'
+      return output.split[1].delete_suffix('!')
+    elsif output.include? 'GitLab'
+      return output.split[3].delete_suffix('!')
```
Comment (bcoles) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: `output.split[3]` may `raise`.

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+      end
+
+      begin
+        1.upto(threads) do
+          t << framework.threads.spawn("Module(#{refname})", false, queue.shift) do |file|
+            Thread.current.kill unless file
+
+            config_contents = "Host gitserver\n\tUser git\n\tHostname #{datastore['GITSERVER']}\n\tPreferredAuthentications publickey\n\tIdentityFile #{file}\n"
+
+            rand_filename = '/tmp/' + Rex::Text.rand_text_alpha(8, bad = '')
```
Comment (bcoles) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: I believe `Rex::Quickfile.new([])` or `Tempfile.new([])` are preferred.

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+      return results
+    end
+
+    check_git_keys(keys)
+  end
+
+  def run
+    if datastore['KEY_FILE'].nil? && datastore['KEY_DIR'].nil?
+      print_error 'Please specify a KEY_FILE or KEY_DIR'
+      return
```
Comment (bcoles) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]:
```ruby
  fail_with Failure::BadConfig, 'Please specify a KEY_FILE or KEY_DIR'
```

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    elsif !(key_file.blank? ^ key_dir.blank?)
+      print_error 'Please only specify one KEY_FILE or KEY_DIR'
+      return
```
Comment (bcoles) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]:
```ruby
  fail_with Failure::BadConfig, 'Please only specify one KEY_FILE or KEY_DIR'
```

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        keys << file unless check_key_for_passphrase(file)
+      end
+    end
+    if keys.empty?
+      print_error "SSH - No valid keys found"
+    end
+    return keys
+  end
+
+  def provide_user(output)
```
Comment (bcoles) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: `provide_user` is not an intuitive method name.

Comment (wdahlenburg) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: Thoughts on parse_user ?

Comment (bcoles) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: Later on, you reject all keys when no `user` is present with `unless user.empty?`. The only place that the `user` is populated is in this method. This implies that this method might do more than extract/parse the username - it is also solely responsible for determining whether authentication was successful.

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb (hunk 3 above)
Comment (wdahlenburg) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: Would matching with regex be a better way to parse this?

Comment (wdahlenburg) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: Switched to regex in d12fce6

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb (hunk 7 above)
Comment (wdahlenburg) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: This was switched to parse_user and the unless statement was removed from the table. Entries will only be added if the username is found and matches the regex per supported git server.

**Comment by wdahlenburg** [comment]
@bcoles @dwelch-r7 @acammack-r7 Do you have any more feedback or reviews?

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+      rescue ::Timeout::Error
+      ensure
+        t.each { |x| x.kill rescue nil }
+      end
+    end
+    return results
+  end
+
+  def test_keys
+    results = {}
```
Comment (dwelch-r7) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: the content of results is never modified or accessed, can just get rid of it

Comment (acammack-r7) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: With the minor exception of https://github.com/rapid7/metasploit-framework/pull/12422/files#diff-25464e4273fbaa35a71e2fbdb2b237f4R150, which would be fine as `return {}`

Comment (dwelch-r7) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: @acammack-r7 yup fair point, should have made that clear 😅

Comment (wdahlenburg) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: Good catch. Updated

File: modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        'SessionTypes'  => ['shell', 'meterpreter'],
+        'References'    => [['URL', 'https://help.github.com/en/articles/testing-your-ssh-connection']]
+      )
+    )
+
+    register_options(
+      [
+        OptPath.new('KEY_FILE', [false, 'Filename of a private key.', nil]),
+        OptPath.new('KEY_DIR', [false, 'Directory of several keys. Filenames will be recursivley found matching id_* (Ex: /home/user/.ssh)', nil]),
+        OptString.new('GITSERVER', [false, 'Optional parameter to specify alternate Git Server (GitHub, GitLab, etc)', 'github.com'])
```
Comment (dwelch-r7) [inline modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: This probably makes more sense as a mandatory option with the default of github (which you already have done) but some sort of git server should be specified

Comment (wdahlenburg) [reply modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb]: 02bb97f should address this

**Comment by dwelch-r7** [comment]
# Release Notes

This module attempts to authenticate with Git servers using provided SSH keys. If successful it will identify the user that the ssh key belongs to.

**Pull request merged**